readthedocs/proxito/middleware.py

```diff
@@ -0,0 +1,127 @@
+    """
+    Take the request and map the host to the proper project slug.
+
+    We check, in order:
+
+    * The ``HTTP_X_RTD_SLUG`` host header for explicit Project mapping
+        - This sets ``request.rtdheader`` True
+    * The ``PUBLIC_DOMAIN`` where we can use the subdomain as the project name
+        - This sets ``request.subdomain`` True
+    * The hostname without port information, which maps to ``Domain`` objects
+        - This sets ``request.cname`` True
+    """
+
+    host = request.get_host().lower().split(':')[0]
+    public_domain = settings.PUBLIC_DOMAIN.lower().split(':')[0]
+    host_parts = host.split('.')
+    public_domain_parts = public_domain.split('.')
+
+    # Explicit Project slug being passed in
+    if 'HTTP_X_RTD_SLUG' in request.META:
+        project_slug = request.META['HTTP_X_RTD_SLUG'].lower()
+        request.rtdheader = True
+
+    elif public_domain in host:
+        # Serve from the PUBLIC_DOMAIN, ensuring it looks like `foo.PUBLIC_DOMAIN`
+        if public_domain_parts == host_parts[1:]:
+            project_slug = host_parts[0]
+            request.subdomain = True
+            log.debug('Proxito Public Domain: host=%s', host)
+        else:
+            # TODO: This can catch some possibly valid domains (docs.readthedocs.io.com) for example
+            # But these feel like they might be phishing, etc. so let's block them for now.
+            project_slug = None
+            log.warning('Weird variation on our hostname: %s', host)
+            return HttpResponseBadRequest(_('Invalid hostname'))
+
+    # Serve CNAMEs
+    else:
+        domain = Domain.objects.filter(domain=host).first()
+        if domain:
+            project_slug = domain.project.slug
+            request.cname = True
+            log.debug('Proxito CNAME: %s', host)
+        else:
+            # Some person is CNAMEing to us without configuring a domain - 404.
+            project_slug = None
+            log.debug('CNAME 404: %s', host)
+            return render(
+                request, 'core/dns-404.html', context={'host': host}, status=404
+            )
+    log.debug('Proxito Project: %s', project_slug)
+    return project_slug
+
+
+class ProxitoMiddleware(MiddlewareMixin):
+
+    """The actual middleware we'll be using in prod."""
+
+    def process_request(self, request):  # noqa
+        if any([not settings.USE_SUBDOMAIN, 'localhost' in request.get_host(),
+                'testserver' in request.get_host()]):
+            log.debug('Not processing Proxito middleware')
+            return None
+
+        ret = map_host_to_project_slug(request)
+
+        # Handle returning a response
+        if hasattr(ret, 'status_code'):
+            return ret
+
+        # Otherwise set the slug on the request
+        request.host_project_slug = request.slug = ret
+
+        return None
+
+
+class NewStyleProxitoMiddleware:
+
+    """The new style middleware, I can't figure out how to test it."""
```

Comment:
I think you can test the request at least by creating a method `process_request(self, request)`, that way you don't depend on `self.get_response` or mock it to test with `localhost` and `testserver` as hostname.

Reply:
And I think you only want to test is the request

Reply:
Yea, the `get_response` thing just feels hard to test. I think having it old-style for now is fine, and we can figure out the migration in the future.

Reply:
Just going to delete the new-style middleware for nw.